damnit/gui/web_viewer.py

```diff
@@ -0,0 +1,87 @@
+from PyQt5 import QtCore, QtWidgets
+from PyQt5.QtCore import QBuffer, QByteArray, QIODevice, QUrl
+from PyQt5.QtGui import QMovie
+from PyQt5.QtWebEngineCore import (QWebEngineUrlRequestJob,
+                                   QWebEngineUrlScheme,
+                                   QWebEngineUrlSchemeHandler)
+from PyQt5.QtWebEngineWidgets import (QWebEnginePage, QWebEngineProfile,
+                                      QWebEngineView)
+
+from ..api import Damnit
+from .widgets import QtWaitingSpinner
+
+LOCAL_SCHEME = QByteArray(b"damnit")
+
+scheme = QWebEngineUrlScheme(LOCAL_SCHEME)
+scheme.setFlags(
+    QWebEngineUrlScheme.Flag.SecureScheme
+    | QWebEngineUrlScheme.Flag.LocalScheme
+    | QWebEngineUrlScheme.Flag.LocalAccessAllowed
+)
+QWebEngineUrlScheme.registerScheme(scheme)
+
+
+class UrlSchemeHandler(QWebEngineUrlSchemeHandler):
+    def install(self, profile):
+        profile.installUrlSchemeHandler(LOCAL_SCHEME, self)
+
+    def requestStarted(self, job):
+        href = job.requestUrl().path()
+        path = href.removeprefix('damn.it')
+        *db_path, proposal, run, name = path.split('/')
+
+        db_path = '/'.join(db_path)
+        if db_path == '':
+            db_path = int(proposal)
+
+        try:
+            _data = Damnit(db_path)[int(run), name].read()
+        except Exception:
+            import traceback
+            traceback.print_exc()
+            print(f"ERROR: request job failed: {href!r}")
+            job.fail(QWebEngineUrlRequestJob.Error.UrlNotFound)
+            return
+
+        mime = QByteArray(b"text/html")
+        buffer = QBuffer(job)
+        buffer.setData(_data.to_html(include_plotlyjs="cdn").encode())
+        buffer.open(QIODevice.OpenModeFlag.ReadOnly)
+        job.reply(mime, buffer)
+
+
+class PlotlyPlot(QtWidgets.QWidget):
+    def __init__(self, variable):
+        super().__init__()
+        self.setGeometry(0, 0, 1024, 768)
+
+        self.url = f"damn.it{variable._db.path.parent}/{variable.proposal}/{variable.run}/{variable.name}"
+
+        self.browser = QWebEngineView(self)
+        profile = QWebEngineProfile.defaultProfile()
```

Comment:
Curious question: what's the difference of the profile that was declared on the `main_window` and the profile that was set here? Would it be possible to reuse it?

Reply:
they are the same. Here I get the [default profile](https://doc.qt.io/qt-6/qwebengineprofile.html#defaultProfile), I don't set a new one.